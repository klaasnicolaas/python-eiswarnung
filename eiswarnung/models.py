"""Data models for the Eiswarnung API."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any


class ForecastType(str, Enum):
    """Enumeration representing the Eiswarnung Forecast type."""

    NO_ICE = "No ice"
    POSSIBLY_ICE = "Possibly ice"
    ICE = "Ice"


@dataclass
class Forecast:
    """Object representing an Forecast response from Eiswarnung."""

    request_date: datetime
    status_id: int
    text: str
    city: str
    forecast_date: datetime

    @property
    def forecast_type(self) -> ForecastType:
        """Return API account_type information.

        Returns:
            The forecast type.
        """
        if self.status_id == 1:
            return ForecastType.ICE
        if self.status_id == 2:
            return ForecastType.POSSIBLY_ICE
        return ForecastType.NO_ICE

    @classmethod
    def from_response(cls, data: dict) -> Forecast:
        """Create an Forecast from a response.

        Args:
            data: The response data from the Eiswarnung API.

        Returns:
            A Forecast object.
        """
        data = data["result"]
        return Forecast(
            request_date=datetime.strptime(
                data.get("requestDate"), "%Y-%m-%d %H:%M:%S"
            ),
            status_id=data.get("forecastId"),
            text=data.get("forecastText").replace(".", ""),
            city=data.get("forecastCity"),
            forecast_date=datetime.strptime(
                data.get("forecastDate"), "%Y-%m-%d"
            ).date(),
        )


@dataclass
class Ratelimit:
    """Information about the current rate limit."""

    call_limit: int
    remaining_calls: int
    retry_after: int

    @classmethod
    def from_response(cls, data: dict[str, Any]) -> Ratelimit:
        """Create a Ratelimit object from the response data.

        Args:
            data: The response data from the Eiswarnung API.

        Returns:
            A Ratelimit object.
        """
        return cls(
            call_limit=data.get("callsDailyLimit"),
            remaining_calls=data.get("callsLeft"),
            retry_after=data.get("callsResetInSeconds"),
        )
