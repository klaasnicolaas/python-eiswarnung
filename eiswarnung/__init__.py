"""Asynchronous client for the Eiswarnung API."""
from .eiswarnung import (
    Eiswarnung,
    EiswarnungConnectionError,
    EiswarnungError,
    EiswarnungRatelimitError,
    EiswarnungRequestError,
)
from .models import Estimate, Ratelimit

__all__ = [
    "Eiswarnung",
    "EiswarnungError",
    "EiswarnungConnectionError",
    "EiswarnungRatelimitError",
    "EiswarnungRequestError",
    "Estimate",
    "Ratelimit",
]
