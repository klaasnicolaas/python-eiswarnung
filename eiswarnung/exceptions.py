"""Exceptions for Eiswarnung API."""


class EiswarnungError(Exception):
    """General Eiswarnung exception."""


class EiswarnungConnectionError(EiswarnungError):
    """Eiswarnung connection exception."""


class EiswarnungRequestError(EiswarnungError):
    """Eiswarnung wrong request input variables."""

    def __init__(self, data: dict) -> None:
        """Init a solar request error.

        Args:
            data: The data that caused the error.

        https://www.eiswarnung.de/rest-api/
        """
        super().__init__(f'{data["message"]} (error {data["code"]})')
        self.code = data["code"]


class EiswarnungRatelimitError(EiswarnungError):
    """Eiswarnung ratelimit exception."""

    def __init__(self, data: dict) -> None:
        """Init a rate limit error.

        Args:
            data: The data that caused the error.
        """
        super().__init__(f'{data["message"]} (error {data["code"]})')
        self.code = data["code"]
