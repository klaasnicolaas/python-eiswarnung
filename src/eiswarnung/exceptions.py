"""Exceptions for Eiswarnung API."""
from typing import Any


class EiswarnungError(Exception):
    """General Eiswarnung exception."""


class EiswarnungConnectionError(EiswarnungError):
    """Eiswarnung connection exception."""


class EiswarnungRequestError(EiswarnungError):
    """Eiswarnung wrong request input variables."""

    def __init__(self, data: dict[str, Any]) -> None:
        """Init a solar request error.

        Args:
        ----
            data: The data that caused the error.

        https://www.eiswarnung.de/rest-api/
        """
        super().__init__(f'{data["message"]} (error {data["code"]})')
        self.code = data["code"]


class EiswarnungConnectionTimeoutError(EiswarnungError):
    """Eiswarnung connection timeout exception."""


class EiswarnungRatelimitError(EiswarnungError):
    """Eiswarnung ratelimit exception."""

    def __init__(self, data: dict[str, Any]) -> None:
        """Init a rate limit error.

        Args:
        ----
            data: The data that caused the error.
        """
        super().__init__(f'{data["message"]} (error {data["code"]})')
        self.code = data["code"]
