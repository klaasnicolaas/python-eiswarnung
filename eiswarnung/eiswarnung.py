"""Asynchronous Python client for the Eiswarnung API."""
from __future__ import annotations

import asyncio
import socket
from collections.abc import Mapping
from dataclasses import dataclass
from importlib import metadata
from typing import Any

import async_timeout
from aiohttp.client import ClientError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import (
    EiswarnungConnectionError,
    EiswarnungError,
    EiswarnungRatelimitError,
    EiswarnungRequestError,
)
from .models import Forecast, Ratelimit


@dataclass
class Eiswarnung:
    """Main class for handling connections with the Eiswarnung API."""

    api_key: str
    latitude: float
    longitude: float

    request_timeout: float = 10.0
    session: ClientSession | None = None
    ratelimit: Ratelimit | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: Mapping[str, str] | None = None,
    ) -> dict[str, Any]:
        """Handle a request to the Eiswarnung API.

        A generic method for sending/handling HTTP requests done against
        the Eiswarnung API.

        Args:
            uri: Request URI, without '/', for example, 'status'.
            method: HTTP method to use.
            params: Dictionary of parameters to send to the Eiswarnung API.

        Returns:
            The response data from the Eiswarnung API.

        Raises:
            EiswarnungRequestError: There is something wrong with the
                variables used in the request.
            EiswarnungConnectionError: An error occurred while communicating
                with the Eiswarnung API.
            EiswarnungError: Received an unexpected response from the
                Eiswarnung API.
            EiswarnungRatelimitError: The number of requests has exceeded
                the rate limit of the Eiswarnung API.
        """
        version = metadata.version(__package__)
        url = URL.build(scheme="https", host="api.eiswarnung.de", path="/").join(
            URL(uri)
        )

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonEiswarnung/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise EiswarnungConnectionError(
                "Timeout occurred while connecting to Eiswarnung API"
            ) from exception
        except (ClientError, socket.gaierror) as exception:
            raise EiswarnungConnectionError(
                "Error occurred while communicating with Eiswarnung API"
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            raise EiswarnungError(
                "Unexpected response from the Eiswarnung API",
                {"Content-Type": content_type, "response": text},
            )

        data = await response.json()
        if data["code"] == 200:
            self.ratelimit = Ratelimit.from_response(data)

        if data["code"] in [300, 400, 401]:
            raise EiswarnungRequestError(data)

        if data["code"] == 402:
            raise EiswarnungRatelimitError(data)

        return data

    async def forecast(self) -> Forecast:
        """Get forecast information from the Eiswarnung API.

        Returns:
            A Forecast object, with a ice warning forecast.
        """
        data = await self._request(
            "/",
            params={
                "key": self.api_key,
                "lat": self.latitude,
                "lng": self.longitude,
            },
        )
        return Forecast.from_response(data)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Eiswarnung:
        """Async enter.

        Returns:
            The Eiswarnung object.
        """
        return self

    async def __aexit__(self, *_exc_info) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
