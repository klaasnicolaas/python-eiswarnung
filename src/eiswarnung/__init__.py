"""Asynchronous client for the Eiswarnung API."""

from .eiswarnung import Eiswarnung
from .exceptions import (
    EiswarnungConnectionError,
    EiswarnungConnectionTimeoutError,
    EiswarnungError,
    EiswarnungRatelimitError,
    EiswarnungRequestError,
)
from .models import Forecast, Ratelimit

__all__ = [
    "Eiswarnung",
    "EiswarnungConnectionError",
    "EiswarnungConnectionTimeoutError",
    "EiswarnungError",
    "EiswarnungRatelimitError",
    "EiswarnungRequestError",
    "Forecast",
    "Ratelimit",
]
