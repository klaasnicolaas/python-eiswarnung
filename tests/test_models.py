"""Test the models."""

from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from eiswarnung import Eiswarnung, Forecast

from . import load_fixtures


async def test_forecast_type0(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    eiswarnung_client: Eiswarnung,
) -> None:
    """Test request for Forecast object."""
    aresponses.add(
        "api.eiswarnung.de",
        "/",
        "GET",
        aresponses.Response(
            text=load_fixtures("forecast_type0.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )
    forecast: Forecast = await eiswarnung_client.forecast()
    assert forecast == snapshot
    assert forecast.forecast_type == "No ice"


async def test_forecast_type1(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    eiswarnung_client: Eiswarnung,
) -> None:
    """Test request for Forecast object."""
    aresponses.add(
        "api.eiswarnung.de",
        "/",
        "GET",
        aresponses.Response(
            text=load_fixtures("forecast_type1.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )
    forecast: Forecast = await eiswarnung_client.forecast()
    assert forecast == snapshot
    assert forecast.forecast_type == "Ice"


async def test_forecast_type2(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    eiswarnung_client: Eiswarnung,
) -> None:
    """Test request for Forecast object."""
    aresponses.add(
        "api.eiswarnung.de",
        "/",
        "GET",
        aresponses.Response(
            text=load_fixtures("forecast_type2.json"),
            status=200,
            headers={"Content-Type": "application/json"},
        ),
    )
    forecast: Forecast = await eiswarnung_client.forecast()
    assert forecast == snapshot
    assert forecast.forecast_type == "Possibly ice"
