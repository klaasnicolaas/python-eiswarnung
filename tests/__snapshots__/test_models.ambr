# serializer version: 1
# name: test_forecast_type0
  Forecast(request_date=datetime.datetime(2021, 12, 27, 15, 28, 7, tzinfo=<DstTzInfo 'CET' LMT+0:18:00 STD>), status_id=0, text='Kein Eis', city='Heidelberg', forecast_date=datetime.date(2021, 12, 28))
# ---
# name: test_forecast_type1
  Forecast(request_date=datetime.datetime(2021, 12, 27, 15, 28, 7, tzinfo=<DstTzInfo 'CET' LMT+0:18:00 STD>), status_id=1, text='Eis', city='Heidelberg', forecast_date=datetime.date(2021, 12, 28))
# ---
# name: test_forecast_type2
  Forecast(request_date=datetime.datetime(2021, 12, 27, 15, 28, 7, tzinfo=<DstTzInfo 'CET' LMT+0:18:00 STD>), status_id=2, text='Evtl Eis', city='Heidelberg', forecast_date=datetime.date(2021, 12, 28))
# ---
